tests/test_references.py

```diff
@@ -186,13 +241,303 @@ async def test_deploy_end_to_end(
+        run_snippet(snippet="refs::plugins::iterates_str_list(['h', 1])")
+    with pytest.raises(PluginTypeException, match="contains a reference"):
+        run_snippet(snippet="refs::plugins::iterates_str_list(['h', refs::create_string_reference('name')])")
+    # Scenario: plugin argument annotated as `Sequence[str | Reference[str]]`
+    run_snippet(snippet="refs::plugins::iterates_str_ref_list(['h', 'e'])")
+    with pytest.raises(PluginTypeException, match=re.escape("Expected type: (Reference[string] | string)[]")):
+        run_snippet(snippet="refs::plugins::iterates_str_ref_list(['h', 1])")
+    run_snippet(snippet="refs::plugins::iterates_str_ref_list(['h', refs::create_string_reference('name')])")
+    # Scenario: plugin argument annotated as `Mapping[str, object]`
+    run_snippet(snippet="refs::plugins::iterates_object_dict({'h': 'h', 'e': 'e'})")
+    with pytest.raises(PluginTypeException, match="contains a reference"):
+        run_snippet(snippet="refs::plugins::iterates_object_dict({'h': 'h', 'e': refs::create_string_reference('name')})")
+    # Scenario: plugin argument annotated as `Mapping[str, object | Reference[object]]`
+    run_snippet(snippet="refs::plugins::iterates_object_ref_dict({'h': 'h', 'e': 'e'})")
+    run_snippet(snippet="refs::plugins::iterates_object_ref_dict({'h': 'h', 'e': refs::create_string_reference('name')})")
+
+    # Entities
+
+    # Scenario: plugin annotated as `Entity`
+    ## Entity annotation
+    ### dataclasses allowed
+    run_snippet("refs::plugins::takes_entity(refs::dc::AllRefsDataclass(maybe_ref_value=refs::create_string_reference('hello')))")
+    run_snippet("refs::plugins::takes_entity(refs::dc::NoRefsDataclass())")
+    ### references allowed, as long as no reference attribute is accessed
+    run_snippet("refs::plugins::takes_entity(refs::dc::create_all_refs_dataclass_reference('hello'))")
+    run_snippet("refs::plugins::takes_entity(refs::dc::create_no_refs_dataclass_reference())")
+
+    # Scenario: plugin annotated as `Entity` accesses reference attribute during plugin execution
+    ## no reference
+    run_snippet("refs::plugins::read_entity_value(refs::dc::AllRefsDataclass(maybe_ref_value='Hello World!'))")
+    run_snippet("refs::plugins::read_entity_list_value(refs::ListContainer(value=['Hello', 'World!']))")
+    run_snippet("refs::plugins::read_entity_list_head(refs::ListContainer(value=['Hello', refs::create_string_reference('hello')]))")
+    run_snippet("refs::plugins::read_entity_dict_value(refs::DictContainer(value={'Hello': 'World!', 'mykey': '42'}))")
+    run_snippet(
+        "refs::plugins::read_entity_dict_mykey(refs::DictContainer(value={'Hello': refs::create_string_reference('hello'), 'mykey': '42'}))"
+    )
+    ## reference
+    ### in attribute
+    with raises_wrapped(UndeclaredReference, match="Encountered undeclared reference .* Encountered at instance.maybe_ref_value"):
```

Comment:
Does  the error provide instructions on what to do when it is allowed?

Reply:
It mentions `... | Reference[...]` type annotations for the boundary, and it points to the `plugins.allow_reference_values()` wrapper for other cases. Would you like me to add that part of the message to the assertions? I left it out mostly for verbosity reasons.